brave/src/main/java/brave/internal/Platform.java

```diff
@@ -34,8 +34,9 @@
  */
 public abstract class Platform {
   private static final Platform PLATFORM = findPlatform();
-  private static final Logger LOG = Logger.getLogger(brave.Tracer.class.getName());
 
+  // Lazy init as a constant eagerly initializes hooks like log4j2 even when the tracer is no-op
```

Comment:
I think we can avoid the volatile read / missing out on potential inlining optimizations by using a nested class

```java
// Use nested class to ensure logger isn't initialized unless it is accessed once.
private static class LoggerHolder {
  private static final Logger LOG = Logger.getLogger(brave.Tracer.class.getName());
}

public void log() {
  LoggerHolder.LOG.log
}
```

Reply:
The same doesn't have to apply to `linkLocalIp` since that's only accessed once when constructing `Tracing` which should be infrequent - maybe it doesn't even need to be memoized because of that actually.

Reply:
awesome idea on the holder thing!

there's a pattern where people create multiple instances of brave, for better or for worse.. I'll leave the linkedlocal ip thing alone for now.